### PR TITLE
cmd/sniproxy: add start of a tsnet-based SNI proxy

### PR DESCRIPTION
**Pull request by bradfitz**
$ curl https://canhazip.com/
    170.173.0.21
    $ curl --resolve canhazip.com:443:100.85.165.81 https://canhazip.com/
    34.223.127.151

Updates #1748

### CODE REVIEW

File: cmd/sniproxy/snipproxy.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		c, err := ln.Accept()
+		if err != nil {
+			log.Fatal(err)
+		}
+		go s.serveConn(c)
+	}
+}
+
+func (s *server) serveConn(c net.Conn) {
+	addrPortStr := c.LocalAddr().String()
```
Comment (maisem) [inline cmd/sniproxy/snipproxy.go:67]: Probably could take this as an input from serve.

**Review by maisem — APPROVED** [review]

File: cmd/sniproxy/snipproxy.go (hunk 1 above)
Comment (bradfitz) [reply cmd/sniproxy/snipproxy.go:67]: I feared there'd be start-up races. I wanted to wire up all that paths before our Tailscale IP is known.

Comment (maisem) [reply cmd/sniproxy/snipproxy.go:67]: I meant just the port

Comment (bradfitz) [reply cmd/sniproxy/snipproxy.go:67]: It can be N ports.

Comment (maisem) [reply cmd/sniproxy/snipproxy.go:67]: this is fine :)

**Pull request merged**